secboot/secboot_sb.go

```diff
@@ -129,6 +130,62 @@ func UnlockVolumeUsingSealedKeyIfEncrypted(disk disks.Disk, name string, sealedE
 	return res, err
 }
 
+func UnlockEncryptedVolumeUsingProtectorKey(disk disks.Disk, name string, key []byte) (UnlockResult, error) {
+	unlockRes := UnlockResult{
+		UnlockMethod: NotUnlocked,
+	}
+
+	// find the encrypted device using the disk we were provided - note that
+	// we do not specify IsDecryptedDevice in opts because here we are
+	// looking for the encrypted device to unlock, later on in the boot
+	// process we will look for the decrypted device to ensure it matches
+	// what we expected
+	part, err := disk.FindMatchingPartitionWithFsLabel(EncryptedPartitionName(name))
+	if err != nil {
+		return unlockRes, err
+	}
+	unlockRes.IsEncrypted = true
+	// we have a device
+	encdev := filepath.Join("/dev/disk/by-uuid", part.FilesystemUUID)
+	unlockRes.PartDevice = encdev
+
+	uuid, err := randutilRandomKernelUUID()
+	if err != nil {
+		// We failed before we could generate the filsystem device path for
+		// the encrypted partition device, so we return FsDevice empty.
+		return unlockRes, err
+	}
+
+	// make up a new name for the mapped device
+	mapperName := name + "-" + uuid
+
+	slots, err := sbListLUKS2ContainerUnlockKeyNames(encdev)
+	if err != nil {
+		return unlockRes, fmt.Errorf("cannot list slots in partition save partition: %v", err)
+	}
+
+	if len(slots) != 0 {
+		const allowRecovery = false
+		options := activateVolOpts(allowRecovery)
+		options.PassphraseTries = 0
+
+		sb_plainkey.SetProtectorKeys(key)
+		defer sb_plainkey.SetProtectorKeys()
+
+		if err := sbActivateVolumeWithKeyData(mapperName, encdev, nil, options); err != nil {
```

Comment:
For the nil we should maybe use a const. This is the auth requester I think.

Reply:
not a bad idea